lib/internal/http2/core.js

```diff
@@ -696,10 +697,14 @@ class Http2Session extends EventEmitter {
 
   // type     { number } either NGHTTP2_SESSION_SERVER or NGHTTP2_SESSION_CLIENT
   // options  { Object }
-  // socket   { net.Socket | tls.TLSSocket }
+  // socket   { net.Socket | tls.TLSSocket | stream.Duplex }
   constructor(type, options, socket) {
     super();
 
+    if (!(socket instanceof net.Socket)) {
```

Comment:
hmm... I'd almost prefer that this would check to see if the `socket` has an `._external` ... that is, rather than checking if it is a `net.Socket`, that it would check if it is a `StreamBase` somehow. Yes, generally it should always be socket but...

Reply:
Also, I've purposefully avoided using `instanceof` in the constructor here to avoid the performance hit. It would be good if we could find a way of avoiding adding it here.

Reply:
Fwiw, for tls we also check `instanceof net.Socket`, although I think I generally agree

Reply:
I’m going to sleep soon but feel free to push to this branch :)

Reply:
I'm working on a few other items so cannot break away to push on this branch.

One thing to keep in mind is that if this mechanism is only going to be used for the client side, the `instanceof` check can be moved into the `ClientHttp2Session` constructor rather than the base `Http2Session` in order to keep the performance hit isolated just to the client side of things.

Reply:
Actually... scratch that... I see you're using this for server side also... hmmm let me stew on it a bit.

Reply:
Okay, I’ve switched to checking for the presence of `socket._handle._externalStream`